src/Microsoft.DotNet.Interactive/Commands/KernelCommand.cs

```diff
@@ -64,4 +134,41 @@ public virtual Task InvokeAsync(KernelInvocationContext context)
 
         return Handler(this, context);
     }
+
+    public void SetId(string id)
+    {
+        _id = id;
+    }
+
+    internal string GetOrCreateId()
+    {
+        if (_id is not null)
+        {
+            return _id;
+        }
+
+        SetId(Guid.NewGuid().ToString("N"));
+
+        return _id;
+    }
+
+    public bool Equals(KernelCommand other)
+    {
+        if (ReferenceEquals(this, other))
+        {
+            return true;
+        }
+
+        if (other is null)
+        {
+            return false;
+        }
+
+        return GetOrCreateId() == other.GetOrCreateId();
+    }
+
+    public override int GetHashCode()
+    {
+        return GetOrCreateId().GetHashCode();
```

Comment:
Shouldn't the hash code be based the token too? Or is this just here temporarily until the id concept is removed?

Reply:
Yeah, so far this part is a refactor with no behavioral change. This will eventually be based on the token when the id concept is gone.